modules/distributor/forwarder/manager.go

```diff
@@ -0,0 +1,370 @@
+package forwarder
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/grafana/dskit/services"
+	"github.com/prometheus/client_golang/prometheus"
+	"go.uber.org/multierr"
+	"golang.org/x/exp/constraints"
+
+	"github.com/grafana/tempo/modules/distributor/queue"
+	"github.com/grafana/tempo/pkg/tempopb"
+)
+
+const (
+	defaultWorkerCount = 2
+	defaultQueueSize   = 100
+)
+
+type Overrides interface {
+	TenantIDs() []string
+	Forwarders(tenantID string) []string
+}
+
+type Manager struct {
+	services.Service
+	logger    log.Logger
+	reg       prometheus.Registerer
+	overrides Overrides
+
+	// forwarderNameToForwarder is static throughout lifecycle of the manager and read-only
+	forwarderNameToForwarder map[string]Forwarder
+
+	tenantToQueueList   map[string]*queueList
+	tenantToQueueListMu *sync.RWMutex
+}
+
+func NewManager(cfgs ConfigList, logger log.Logger, reg prometheus.Registerer, overrides Overrides) (*Manager, error) {
+	if err := cfgs.Validate(); err != nil {
+		return nil, fmt.Errorf("failed to validate config list: %w", err)
+	}
+
+	forwarderNameToForwarder := make(map[string]Forwarder, len(cfgs))
+	for i, cfg := range cfgs {
+		forwarder, err := New(cfg, logger)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create forwarder for cfg at index=%d: %w", i, err)
+		}
+
+		forwarderNameToForwarder[cfg.Name] = forwarder
+	}
+
+	m := &Manager{
+		logger:                   logger,
+		reg:                      reg,
+		overrides:                overrides,
+		forwarderNameToForwarder: forwarderNameToForwarder,
+		tenantToQueueList:        make(map[string]*queueList),
+		tenantToQueueListMu:      &sync.RWMutex{},
+	}
+
+	m.Service = services.NewBasicService(m.start, m.run, m.stop)
+
+	return m, nil
+}
+
+func (m *Manager) ForTenant(tenantID string) List {
+	list, found := m.getQueueList(tenantID)
+	if !found {
+		return nil
+	}
+
+	return list.copy()
```

Comment:
We cannot return the original slice. First of all, there is no "original" slice. `list.copy()` constructs a slice from a map. Secondly, the underlying map can (and will) change on overrides configuration updates.

If we want to get rid of those allocations, perhaps the `ForTenant()` could be changed to `ForwardBatchesForTenant()` and we would just call forward batches on each queue under a lock.

Let me know your thoughts.

Reply:
Making the queueList immutable as described below opens up the possibility to return a slice owned by the queueList for reduced allocations. I'll admit I'm not honestly sure what the impact will be here. It will certainly be far less then the proto marshalling discussion below.